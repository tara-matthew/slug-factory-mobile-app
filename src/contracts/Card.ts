export interface ICardProps {
    item: {
        title: string;
    };
    imageURL: string;
    blurhash: string;
    sendDataToParent: (data: never) => void;
}
