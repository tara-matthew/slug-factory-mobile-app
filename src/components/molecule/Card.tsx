import React, { memo } from "react";
import { Text, StyleSheet, View, Pressable } from "react-native";
import { ICardProps } from "../../contracts/Card";
import { Image } from "expo-image";

const Card = ({ item, imageURL, blurhash, sendDataToParent }: ICardProps) => {
    const height = 200;

    // TODO add a variant for full width vs other width?

    const uri = imageURL.startsWith("https") ? imageURL : `${process.env.EXPO_PUBLIC_URL}/${imageURL}`;
    const buttonPressed = (item: { title: string }) => {
        sendDataToParent(item);
    };

    return (
        <View style={ [styles.container] }>
            <Pressable
                accessibilityRole="button"
                onPress={ () => buttonPressed(item) }
            >
                <View className="relative">
                    <Image
                        placeholder={ { blurhash } }
                        transition={ 1000 }
                        style={ {
                            width: "100%",
                            height: height,
                            borderTopLeftRadius: 8,
                            borderTopRightRadius: 8,
                        } }
                        source={ {
                            uri: uri,
                        } }
                    />
                    <View style={ { height: 50, flex: 1, justifyContent: "center", alignItems: "center", paddingHorizontal: 8, paddingVertical: 8 } }>
                        <Text
                            adjustsFontSizeToFit={ false }
                            numberOfLines={ 1 }
                            style={ { fontSize: 18 } }
                        >
                            {item.title}
                        </Text>
                        {item.extraData && (
                            <Text adjustsFontSizeToFit={ false } numberOfLines={ 1 } style={ { fontSize: 13 } }>{item.extraData}</Text>
                        )}
                    </View>
                </View>
            </Pressable>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 0.5,
        backgroundColor: "#d0cadb",
        marginBottom: 20,
        shadowRadius: 2,
        borderRadius: 8,
        // maxWidth: "50%",
    },
    icon: {
        textShadowColor: "rgba(0, 0, 0, 0.75)",
        textShadowOffset: { width: -1, height: 1 },
        textShadowRadius: 3,
    },
});

export default memo(Card);
