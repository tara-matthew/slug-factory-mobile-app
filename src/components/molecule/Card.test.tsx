import React from "react";
import Card from "./Card";
import { render, fireEvent, screen } from "@testing-library/react-native";

it("renders correctly", () => {
    const mockSendDataToParent = jest.fn();
    const blurhash
        = "|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[";
    const item
        = {
            title: "A card",

            images: [
                {
                    id: 1,
                    url: "test",
                },
            ],
        };

    const { toJSON } = render(<Card item={ item } blurhash={ blurhash } sendDataToParent={ mockSendDataToParent } imageURL="https://test.com" />);

    expect(toJSON()).toMatchSnapshot();
});
it("displays text", () => {
    const item
        = {
            title: "A card",
            images: [
                {
                    id: 1,
                    url: "test-url.com'",
                },
            ],
        };

    const mockSendDataToParent = jest.fn();
    const blurhash
        = "|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[";

    const { getByText } = render(<Card item={ item } blurhash={ blurhash } imageURL="https://test-url.com" sendDataToParent={ mockSendDataToParent } />);

    expect(getByText("A card")).toBeTruthy();
});

it("sends the item data when the card is pressed", () => {
    const item
        = {
            title: "A card",
            images: [
                {
                    id: 1,
                    url: "test-url.com'",
                },
            ],
        };

    const mockSendDataToParent = jest.fn();
    const blurhash
        = "|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[";

    render(<Card item={ item } blurhash={ blurhash } imageURL="https://test-url.com" sendDataToParent={ mockSendDataToParent } />);

    const button = screen.getByRole("button");
    fireEvent.press(button);
    expect(mockSendDataToParent).toHaveBeenCalledTimes(1);
    expect(mockSendDataToParent).toHaveBeenCalledWith(item);
});
