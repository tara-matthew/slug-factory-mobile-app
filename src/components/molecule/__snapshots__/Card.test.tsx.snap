// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#d0cadb",
        "flex": 1,
        "marginBottom": 20,
        "shadowRadius": 2,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      className="relative"
    >
      <Image
        placeholder={
          {
            "blurhash": "|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[",
          }
        }
        source={
          {
            "uri": "https://test.com",
          }
        }
        style={
          {
            "height": 200,
            "width": "100%",
          }
        }
        transition={1000}
      />
      Icon: favorite-outline, Size: 32, Color: white
      <Text
        adjustsFontSizeToFit={true}
        numberOfLines={2}
        style={
          {
            "fontSize": 18,
            "height": 50,
            "padding": 8,
          }
        }
      >
        A card
      </Text>
    </View>
  </View>
</View>
`;
